File: src/lib/moteur/agenda.ts
/**
 * agenda.ts — Placement des séances et des repas dans la journée réelle.
 *
 * Contraintes appliquées :
 *  - pas de séance intense dans les 90 min précédant le coucher ;
 *  - marge de douche avant le départ au travail pour les séances du matin ;
 *  - délais digestifs autour des repas ;
 *  - un repas de récupération après une séance tardive.
 */

import type { Derive, Jour, Profil, Repas } from "./types";
import { JOURS } from "./types";
import { MINUTE_JOUR, contrainte, fmt, toMin } from "./noyau";

export interface Creneau {
  debut: number;
  fin: number;
  jour: Jour;
}

/** Douche + habillage après une séance matinale, avant de partir. */
export const MARGE_DOUCHE = 25;

function soustraire(creneaux: Creneau[], deb: number, fin: number): Creneau[] {
  const out: Creneau[] = [];
  for (const c of creneaux) {
    if (c.fin <= deb || fin <= c.debut) {
      out.push(c);
      continue;
    }
    if (c.debut < deb) out.push({ debut: c.debut, fin: deb, jour: c.jour });
    if (fin < c.fin) out.push({ debut: fin, fin: c.fin, jour: c.jour });
  }
  return out;
}

/** Créneaux réellement disponibles un jour donné. */
export function creneauxLibres(p: Profil, d: Derive, jour: Jour): Creneau[] {
  const reveil = toMin(p.heureReveil);
  let coucher = toMin(p.heureCoucher);
  if (coucher <= reveil) coucher += MINUTE_JOUR;

  let base: Creneau[] = [{ debut: reveil + 20, fin: coucher - 30, jour }];
  if (base[0].fin <= base[0].debut) return [];

  if (p.joursTravailles.includes(jour)) {
    let fin = d.retourDomicile;
    if (fin <= d.departDomicile) fin += MINUTE_JOUR;
    base = soustraire(base, d.departDomicile, fin);
  }

  for (const i of p.indisponibilites ?? []) {
    const jours = i.jours?.length ? i.jours : JOURS;
    if (!jours.includes(jour)) continue;
    const deb = toMin(i.debut);
    let fin = toMin(i.fin);
    if (fin <= deb) fin += MINUTE_JOUR;
    base = soustraire(base, deb, fin);
  }

  return base.filter((c) => c.fin - c.debut >= 15);
}

function scoreJour(p: Profil, d: Derive, jour: Jour, dureeCible: number): number {
  const cr = creneauxLibres(p, d, jour);
  if (!cr.length) return 0;
  const meilleur = Math.max(...cr.map((c) => c.fin - c.debut));
  if (meilleur < dureeCible * 0.6) return 0;
  const total = cr.reduce((a, c) => a + (c.fin - c.debut), 0);
  let s = Math.min(1, meilleur / dureeCible) * 60 + Math.min(total, 480) / 24;
  if (!p.joursTravailles.includes(jour)) s += 15;
  return s;
}

/** Répartitions privilégiant l'alternance effort / récupération. */
const REPARTITIONS: Record<number, Jour[]> = {
  1: ["samedi"],
  2: ["mardi", "samedi"],
  3: ["lundi", "mercredi", "vendredi"],
  4: ["lundi", "mardi", "jeudi", "vendredi"],
  5: ["lundi", "mardi", "mercredi", "vendredi", "samedi"],
  6: ["lundi", "mardi", "mercredi", "vendredi", "samedi", "dimanche"],
  7: [...JOURS],
};

export function joursEntrainement(p: Profil, d: Derive, dureeCible: number): Jour[] {
  const n = p.seancesParSemaine;
  if (n <= 0) return [];
  if (n >= 7) return [...JOURS];

  const modele = REPARTITIONS[n] ?? REPARTITIONS[3];
  const scores = Object.fromEntries(
    JOURS.map((j) => [j, scoreJour(p, d, j, dureeCible)]),
  ) as Record<Jour, number>;

  const retenus = modele.filter((j) => scores[j] > 0);
  const manquants = n - retenus.length;
  if (manquants > 0) {
    JOURS.filter((j) => !retenus.includes(j))
      .sort((a, b) => scores[b] - scores[a])
      .slice(0, manquants)
      .forEach((j) => {
        if (scores[j] > 0) retenus.push(j);
      });
  }
  return Array.from(new Set(retenus)).sort(
    (a, b) => JOURS.indexOf(a) - JOURS.indexOf(b),
  );
}

export function momentPrefere(p: Profil, d: Derive): "matin" | "midi" | "soir" {
  if (d.fenetreMatin >= 75) return "matin";
  if (d.fenetreSoir >= 75) return "soir";
  return d.fenetreMatin >= d.fenetreSoir ? "matin" : "soir";
}

export function placerSeance(
  p: Profil, d: Derive, jour: Jour, duree: number,
  intensite: "moderee" | "elevee" = "moderee",
): Creneau | null {
  let bruts = creneauxLibres(p, d, jour);

  // Réserve la marge de douche sur le créneau du matin des jours travaillés
  if (p.joursTravailles.includes(jour)) {
    bruts = bruts
      .map((c) =>
        c.fin >= d.departDomicile && d.departDomicile > c.debut
          ? { ...c, fin: d.departDomicile - MARGE_DOUCHE }
          : c,
      )
      .filter((c) => c.fin - c.debut >= 15);
  }

  const cr = bruts.filter((c) => c.fin - c.debut >= duree);
  if (!cr.length) return null;

  let coucher = toMin(p.heureCoucher);
  if (coucher <= toMin(p.heureReveil)) coucher += MINUTE_JOUR;
  const limiteIntense = coucher - 90;

  const pref = momentPrefere(p, d);
  const ancre =
    pref === "matin" ? toMin(p.heureReveil) + 60
      : pref === "midi" ? 750
        : d.retourDomicile + 30;

  let meilleur: Creneau | null = null;
  let meilleurCout: [number, number, number] | null = null;

  for (const c of cr) {
    const depart = Math.max(c.debut, Math.min(ancre, c.fin - duree));
    const fin = depart + duree;
    const pen = intensite === "elevee" && fin > limiteIntense ? fin - limiteIntense : 0;
    const cout: [number, number, number] = [pen, Math.abs(depart - ancre), -(c.fin - c.debut)];
    if (
      !meilleurCout ||
      cout[0] < meilleurCout[0] ||
      (cout[0] === meilleurCout[0] && cout[1] < meilleurCout[1]) ||
      (cout[0] === meilleurCout[0] && cout[1] === meilleurCout[1] && cout[2] < meilleurCout[2])
    ) {
      meilleurCout = cout;
      meilleur = c;
    }
  }
  if (!meilleur) return null;
  const depart = Math.max(meilleur.debut, Math.min(ancre, meilleur.fin - duree));
  return { debut: depart, fin: depart + duree, jour };
}

/** Essaie la durée idéale, puis raccourcit par paliers de 5 min. */
export function placerAvecRepli(
  p: Profil, d: Derive, jour: Jour, dureeIdeale: number,
  dureeMin: number, intensite: "moderee" | "elevee",
): { creneau: Creneau | null; duree: number } {
  let duree = dureeIdeale;
  while (duree >= dureeMin) {
    const c = placerSeance(p, d, jour, duree, intensite);
    if (c) return { creneau: c, duree };
    duree -= 5;
  }
  return { creneau: null, duree: 0 };
}

type RepasBrut = Omit<Repas, "kcal" | "proteinesG" | "glucidesG" | "lipidesG">;

/** Positionne les repas selon l'agenda et l'heure de la séance. */
export function structureRepas(
  p: Profil, d: Derive, jour: Jour, seance: Creneau | null,
): RepasBrut[] {
  const reveil = toMin(p.heureReveil);
  let coucher = toMin(p.heureCoucher);
  if (coucher <= reveil) coucher += MINUTE_JOUR;

  const travaille = p.joursTravailles.includes(jour);
  const repas: RepasBrut[] = [];

  const sautPdj = contrainte(p, "faible_appetit_matin") && d.fenetreMatin < 45;
  const seanceMatin = !!seance && seance.debut < 660;

  if (sautPdj) {
    repas.push({
      nom: "Collation matinale (nomade)", heure: reveil + 120, heureTxt: "",
      duree: 10, lieu: "nomade", role: "appoint",
      note: "Appétit faible au réveil : repoussé et allégé.",
    });
  } else {
    let hPdj = reveil + 30;
    if (travaille && hPdj + 15 > d.departDomicile) {
      hPdj = Math.max(reveil + 15, d.departDomicile - 20);
    }
    let nomPdj = "Petit-déjeuner";
    let notePdj: string | undefined;

    if (seanceMatin && seance) {
      const limite = travaille ? d.departDomicile - 10 : coucher;
      if (seance.debut - (hPdj + 20) >= 75) {
        // Digestion suffisante : on ne touche à rien.
      } else if (seance.fin + 20 <= limite) {
        // On scinde : énergie rapide avant, vrai repas après l'effort.
        repas.push({
          nom: "Pré-séance léger", heure: Math.max(reveil + 10, seance.debut - 35),
          heureTxt: "", duree: 8, lieu: "domicile", role: "pre_effort",
          note: "Séance matinale : banane ou dattes + café, digestion rapide.",
        });
        hPdj = seance.fin + 15;
      } else {
        hPdj = Math.max(reveil + 10, seance.debut - 50);
        nomPdj = "Petit-déjeuner allégé (pré-séance)";
        notePdj = "Moins de 1 h avant l'effort : glucides simples, peu de gras et de fibres. "
          + "Compléter les protéines au déjeuner.";
      }
    }

    repas.push({
      nom: nomPdj, heure: hPdj, heureTxt: "",
      duree: d.fenetreMatin < 60 ? 15 : 25,
      lieu: travaille && hPdj > d.departDomicile ? "nomade" : "domicile",
      role: "demarrage", note: notePdj,
    });
  }

  let hDej: number;
  let lieuDej: string;
  if (travaille) {
    const milieu = Math.floor((d.departDomicile + d.retourDomicile) / 2);
    hDej = Math.max(705, Math.min(810, milieu));
    lieuDej = p.lieuRepas;
  } else {
    hDej = 750;
    lieuDej = "domicile";
  }
  repas.push({ nom: "Déjeuner", heure: hDej, heureTxt: "", duree: 30, lieu: lieuDej, role: "principal" });

  let hDiner = Math.max(d.retourDomicile + 30, 1140);
  hDiner = Math.min(hDiner, coucher - 90);
  if (seance && seance.fin > hDiner - 30) {
    hDiner = Math.max(hDiner, Math.min(seance.fin + 30, coucher - 60));
  }
  repas.push({ nom: "Dîner", heure: hDiner, heureTxt: "", duree: 30, lieu: "domicile", role: "principal" });

  const ecart = hDiner - hDej;
  if (ecart > 330) {
    repas.push({
      nom: "Goûter", heure: hDej + Math.floor(ecart / 2), heureTxt: "",
      duree: 10, lieu: travaille ? "bureau" : "domicile", role: "appoint",
    });
  }

  if (seance) {
    const precedents = repas.filter((r) => r.heure <= seance.debut);
    if (precedents.length) {
      const dernier = precedents.reduce((a, b) => (a.heure > b.heure ? a : b));
      if (seance.debut - dernier.heure > 210) {
        repas.push({
          nom: "Collation pré-séance", heure: seance.debut - 60, heureTxt: "",
          duree: 8, lieu: "nomade", role: "pre_effort",
          note: "Glucides rapides, peu de fibres et de gras.",
        });
      }
    }
    const posterieurs = repas.filter((r) => r.heure >= seance.fin);
    if (!posterieurs.length && seance.fin < coucher - 45) {
      repas.push({
        nom: "Collation de récupération", heure: Math.min(seance.fin + 25, coucher - 30),
        heureTxt: "", duree: 10, lieu: "domicile", role: "post_effort",
        note: "Protéines et glucides dans les 90 min suivant l'effort.",
      });
    }
  }

  repas.sort((a, b) => a.heure - b.heure);
  return repas.map((r) => ({ ...r, heureTxt: fmt(r.heure) }));
}

export function budgetTempsCuisine(p: Profil, d: Derive) {
  const t = p.tempsCuisine;
  const batch = t < 45 || d.pressionTemporelle === "critique" || d.pressionTemporelle === "forte";
  const alaise = ["moyen", "bon", "chef"].includes(p.niveauCuisine);
  return {
    totalJour: t,
    batchCooking: batch,
    sessionBatch: batch ? (alaise ? 90 : 60) : 0,
    jourBatch: d.joursRepos.includes("dimanche") ? "dimanche" : (d.joursRepos[0] ?? "dimanche"),
    cuisineParRepas: Math.max(5, Math.floor(t / 3)),
    strategie: batch ? "batch cooking + assemblage rapide" : "cuisine quotidienne à la demande",
  };
}

File: moteur-python/moteur/agenda.py
"""
agenda.py — Placement temporel des séances et des repas.

Rôle : transformer (réveil, coucher, travail, trajet, indisponibilités) en
créneaux libres réels, puis y caser les séances d'entraînement et les repas
selon des règles de chronobiologie et de digestion.

Contraintes physiologiques appliquées :
  - pas de séance intense dans les 90 min précédant le coucher (sommeil)
  - pas de séance intense dans les 60 min suivant le réveil sans échauffement long
  - délai digestif : 2 h après un repas complet, 30-45 min après une collation
  - un repas dans les 90 min post-séance (fenêtre de récupération)
"""

from __future__ import annotations
from dataclasses import dataclass
from .profil import Profil, JOURS, to_min, to_time, fmt, duree_fmt, Plage

MINUTE_JOUR = 24 * 60


# --------------------------------------------------------------------------
# Créneaux
# --------------------------------------------------------------------------

@dataclass
class Creneau:
    debut: int          # minutes depuis minuit
    fin: int
    jour: str
    etiquette: str = "libre"

    @property
    def duree(self) -> int:
        return self.fin - self.debut

    def chevauche(self, autre: "Creneau") -> bool:
        return self.jour == autre.jour and self.debut < autre.fin and autre.debut < self.fin

    def __str__(self) -> str:
        return f"{self.jour} {fmt(self.debut)}–{fmt(self.fin)} ({duree_fmt(self.duree)}) {self.etiquette}"


def soustraire(creneaux: list[Creneau], bloc: Creneau) -> list[Creneau]:
    """Retire `bloc` d'une liste de créneaux (découpe si nécessaire)."""
    out = []
    for c in creneaux:
        if c.jour != bloc.jour or not c.chevauche(bloc):
            out.append(c)
            continue
        if c.debut < bloc.debut:
            out.append(Creneau(c.debut, bloc.debut, c.jour, c.etiquette))
        if bloc.fin < c.fin:
            out.append(Creneau(bloc.fin, c.fin, c.jour, c.etiquette))
    return out


def creneaux_libres(p: Profil, jour: str, marge_reveil=20, marge_coucher=30) -> list[Creneau]:
    """
    Créneaux réellement disponibles un jour donné, une fois retirés :
    sommeil, travail, trajets, indisponibilités déclarées, et les marges
    incompressibles (toilette matin, routine du soir).
    """
    reveil = to_min(p.heure_reveil)
    coucher = to_min(p.heure_coucher)
    if coucher <= reveil:
        coucher += MINUTE_JOUR

    base = [Creneau(reveil + marge_reveil, coucher - marge_coucher, jour, "libre")]
    if base[0].duree <= 0:
        return []

    if jour in p.jours_travailles:
        # trajet aller + travail + trajet retour = un seul bloc occupé
        bloc = Creneau(p.depart_domicile_min, p.retour_domicile_min, jour, "travail")
        if bloc.fin <= bloc.debut:
            bloc.fin += MINUTE_JOUR
        base = soustraire(base, bloc)

    for ind in p.indisponibilites:
        if ind.concerne(jour):
            base = soustraire(base, Creneau(ind.debut_min, ind.fin_min, jour, ind.motif))

    return [c for c in base if c.duree >= 15]


def tous_creneaux(p: Profil) -> dict[str, list[Creneau]]:
    return {j: creneaux_libres(p, j) for j in JOURS}


# --------------------------------------------------------------------------
# Choix des jours d'entraînement
# --------------------------------------------------------------------------

# Répartitions hebdomadaires privilégiant la récupération (jamais 3 jours
# consécutifs de force en dessous de 5 séances/semaine).
REPARTITIONS = {
    1: ["samedi"],
    2: ["mardi", "samedi"],
    3: ["lundi", "mercredi", "vendredi"],
    4: ["lundi", "mardi", "jeudi", "vendredi"],
    5: ["lundi", "mardi", "mercredi", "vendredi", "samedi"],
    6: ["lundi", "mardi", "mercredi", "vendredi", "samedi", "dimanche"],
    7: list(JOURS),
}


def score_jour(p: Profil, jour: str, duree_cible: int) -> float:
    """
    Qualité d'un jour pour s'entraîner : temps disponible, continuité du
    créneau, et bonus si c'est un jour de repos professionnel.
    """
    cr = creneaux_libres(p, jour)
    if not cr:
        return 0.0
    meilleur = max(c.duree for c in cr)
    if meilleur < duree_cible * 0.6:
        return 0.0
    total = sum(c.duree for c in cr)
    s = min(1.0, meilleur / duree_cible) * 60 + min(total, 480) / 24
    if jour not in p.jours_travailles:
        s += 15
    return s


def jours_entrainement(p: Profil, duree_cible: int) -> list[str]:
    """
    Sélectionne les N meilleurs jours en respectant, autant que possible,
    l'alternance effort/récupération.
    """
    n = p.seances_par_semaine
    if n <= 0:
        return []
    if n >= 7:
        return list(JOURS)

    modele = REPARTITIONS.get(n, REPARTITIONS[3])
    scores = {j: score_jour(p, j, duree_cible) for j in JOURS}

    # 1) on part du modèle théorique, on remplace les jours impraticables
    retenus = [j for j in modele if scores[j] > 0]
    manquants = n - len(retenus)
    if manquants > 0:
        reste = sorted((j for j in JOURS if j not in retenus),
                       key=lambda j: -scores[j])
        for j in reste[:manquants]:
            if scores[j] > 0:
                retenus.append(j)

    # 2) si toujours incomplet, l'agenda est saturé : on renvoie ce qu'on a
    return sorted(set(retenus), key=lambda j: JOURS.index(j))


# --------------------------------------------------------------------------
# Placement d'une séance dans la journée
# --------------------------------------------------------------------------

def moment_prefere(p: Profil) -> str:
    """'matin' | 'midi' | 'soir' — d'après la géométrie de l'agenda."""
    if p.fenetre_matin_min >= 75:
        return "matin"
    if p.fenetre_soir_min >= 75:
        return "soir"
    if p.fenetre_matin_min >= p.fenetre_soir_min:
        return "matin"
    return "soir"


MARGE_DOUCHE = 25   # douche + habillage après une séance, avant de partir


def placer_seance(p: Profil, jour: str, duree: int,
                  intensite: str = "moderee") -> Creneau | None:
    """
    Place une séance de `duree` minutes dans le meilleur créneau du jour.

    Règles :
      - séance intense interdite dans les 90 min avant le coucher
      - une séance du matin en jour travaillé doit laisser MARGE_DOUCHE
        minutes avant le départ au travail
      - on privilégie le créneau le plus long, puis le moment préféré
      - si aucun créneau ne tient la durée, renvoie None (l'appelant
        réduira la durée : voir `placer_avec_repli`)
    """
    bruts = creneaux_libres(p, jour)

    # Réserve de la marge douche sur le créneau du matin des jours travaillés
    if jour in p.jours_travailles:
        depart = p.depart_domicile_min
        ajustes = []
        for c in bruts:
            if c.fin >= depart > c.debut:
                c = Creneau(c.debut, depart - MARGE_DOUCHE, c.jour, c.etiquette)
            if c.duree >= 15:
                ajustes.append(c)
        bruts = ajustes

    cr = [c for c in bruts if c.duree >= duree]
    if not cr:
        return None

    coucher = to_min(p.heure_coucher)
    if coucher <= to_min(p.heure_reveil):
        coucher += MINUTE_JOUR
    limite_intense = coucher - 90

    pref = moment_prefere(p)
    ancre = {"matin": to_min(p.heure_reveil) + 60,
             "midi": 12 * 60 + 30,
             "soir": p.retour_domicile_min + 30}[pref]

    def cout(c: Creneau) -> tuple:
        # départ au plus tôt dans le créneau, mais pas avant l'ancre
        depart = max(c.debut, min(ancre, c.fin - duree))
        fin = depart + duree
        penalite_sommeil = 0
        if intensite in ("elevee", "maximale") and fin > limite_intense:
            penalite_sommeil = (fin - limite_intense)
        return (penalite_sommeil, abs(depart - ancre), -c.duree)

    meilleur = min(cr, key=cout)
    depart = max(meilleur.debut, min(ancre, meilleur.fin - duree))
    return Creneau(depart, depart + duree, jour, "entrainement")


def placer_avec_repli(p: Profil, jour: str, duree_ideale: int,
                      duree_min: int = 20, intensite="moderee") -> tuple[Creneau | None, int]:
    """
    Essaie la durée idéale, puis raccourcit par paliers de 5 min jusqu'à
    `duree_min`. Renvoie (créneau, durée retenue).
    """
    d = duree_ideale
    while d >= duree_min:
        c = placer_seance(p, jour, d, intensite)
        if c:
            return c, d
        d -= 5
    return None, 0


# --------------------------------------------------------------------------
# Repas
# --------------------------------------------------------------------------

def structure_repas(p: Profil, jour: str, seance: Creneau | None = None) -> list[dict]:
    """
    Positionne petit-déjeuner, déjeuner, dîner et collations en fonction de
    l'agenda réel et de l'heure de la séance.

    Renvoie une liste de dicts : {nom, heure, duree, lieu, role}
    """
    reveil = to_min(p.heure_reveil)
    coucher = to_min(p.heure_coucher)
    if coucher <= reveil:
        coucher += MINUTE_JOUR

    travaille = jour in p.jours_travailles
    repas: list[dict] = []

    # --- Petit-déjeuner ---
    saut_pdj = p.contrainte("faible_appetit_matin") and p.fenetre_matin_min < 45
    seance_matin = seance is not None and seance.debut < 11 * 60

    if saut_pdj:
        repas.append({"nom": "Collation matinale (nomade)", "heure": reveil + 120,
                      "duree": 10, "lieu": "nomade", "role": "appoint",
                      "note": "Appétit faible au réveil : repoussé et allégé."})
    else:
        h_pdj = reveil + 30
        if travaille and h_pdj + 15 > p.depart_domicile_min:
            h_pdj = max(reveil + 15, p.depart_domicile_min - 20)

        nom_pdj, note_pdj = "Petit-déjeuner", None

        if seance_matin:
            limite_pdj = (p.depart_domicile_min - 10) if travaille else coucher
            if seance.debut - (h_pdj + 20) >= 75:
                pass  # cas idéal : le petit-déjeuner est digéré avant l'effort
            elif seance.fin + 20 <= limite_pdj:
                # On scinde : gorgée d'énergie avant, vrai repas après l'effort.
                repas.append({"nom": "Pré-séance léger", "heure": max(reveil + 10, seance.debut - 35),
                              "duree": 8, "lieu": "domicile", "role": "pre_effort",
                              "note": "Séance matinale : banane/dattes + café, digestion rapide."})
                h_pdj = seance.fin + 15
            else:
                # Agenda trop serré pour scinder : on allège le repas sur place.
                h_pdj = max(reveil + 10, seance.debut - 50)
                nom_pdj = "Petit-déjeuner allégé (pré-séance)"
                note_pdj = ("Moins de 1 h avant l'effort : glucides simples, "
                            "peu de gras et de fibres. Compléter les protéines au déjeuner.")

        entree = {"nom": nom_pdj, "heure": h_pdj,
                  "duree": 15 if p.fenetre_matin_min < 60 else 25,
                  "lieu": "nomade" if (travaille and h_pdj > p.depart_domicile_min) else "domicile",
                  "role": "demarrage"}
        if note_pdj:
            entree["note"] = note_pdj
        repas.append(entree)

    # --- Déjeuner ---
    if travaille:
        milieu = (p.depart_domicile_min + p.retour_domicile_min) // 2
        h_dej = max(11 * 60 + 45, min(13 * 60 + 30, milieu))
        lieu_dej = p.lieu_repas if p.lieu_repas != "domicile" else "domicile"
    else:
        h_dej = 12 * 60 + 30
        lieu_dej = "domicile"
    repas.append({"nom": "Déjeuner", "heure": h_dej, "duree": 30,
                  "lieu": lieu_dej, "role": "principal"})

    # --- Dîner ---
    h_diner = max(p.retour_domicile_min + 30, 19 * 60)
    h_diner = min(h_diner, coucher - 90)          # 1h30 mini avant le coucher
    if seance and seance.fin > h_diner - 30:      # séance tardive : on décale
        h_diner = max(h_diner, min(seance.fin + 30, coucher - 60))
    repas.append({"nom": "Dîner", "heure": h_diner, "duree": 30,
                  "lieu": "domicile", "role": "principal"})

    # --- Collations conditionnelles ---
    ecart_dej_diner = h_diner - h_dej
    if ecart_dej_diner > 330:  # plus de 5h30 entre les deux
        h_col = h_dej + ecart_dej_diner // 2
        repas.append({"nom": "Goûter", "heure": h_col, "duree": 10,
                      "lieu": "bureau" if travaille else "domicile", "role": "appoint"})

    if seance:
        # collation pré-séance si le dernier repas est loin
        precedents = [r for r in repas if r["heure"] <= seance.debut]
        if precedents:
            dernier = max(precedents, key=lambda r: r["heure"])
            if seance.debut - dernier["heure"] > 210:
                repas.append({"nom": "Collation pré-séance", "heure": seance.debut - 60,
                              "duree": 8, "lieu": "nomade", "role": "pre_effort",
                              "note": "Glucides rapides, peu de fibres et de gras."})
        # repas de récupération si la séance finit tard après le dîner
        posterieurs = [r for r in repas if r["heure"] >= seance.fin]
        if not posterieurs and seance.fin < coucher - 45:
            repas.append({"nom": "Collation de récupération", "heure": min(seance.fin + 25, coucher - 30),
                          "duree": 10, "lieu": "domicile", "role": "post_effort",
                          "note": "Protéines + glucides dans les 90 min post-effort."})

    repas.sort(key=lambda r: r["heure"])
    for r in repas:
        r["heure_txt"] = fmt(r["heure"])
    return repas


def conflit_digestion(repas: list[dict], seance: Creneau | None) -> list[str]:
    """Alertes si un repas complet tombe trop près d'une séance."""
    if not seance:
        return []
    alertes = []
    for r in repas:
        if r["role"] != "principal":
            continue
        fin_repas = r["heure"] + r["duree"]
        if 0 <= seance.debut - fin_repas < 90:
            alertes.append(
                f"{r['nom']} se termine {seance.debut - fin_repas} min avant la séance : "
                f"alléger (moins de gras/fibres) ou décaler de 30 min.")
        if 0 <= r["heure"] - seance.fin < 20:
            alertes.append(f"{r['nom']} démarre juste après la séance : prévoir 15 min de retour au calme.")
    return alertes


def budget_temps_cuisine(p: Profil) -> dict:
    """
    Répartit le temps de cuisine quotidien déclaré et décide si le batch
    cooking est nécessaire.
    """
    t = p.temps_cuisine_min
    batch = t < 45 or p.pression_temporelle in ("critique", "forte")
    return {
        "total_jour_min": t,
        "batch_cooking": batch,
        "session_batch_min": 90 if batch and p.niveau_cuisine in ("moyen", "bon", "chef") else (60 if batch else 0),
        "jour_batch": "dimanche" if "dimanche" in p.jours_repos_semaine else (p.jours_repos_semaine[0] if p.jours_repos_semaine else "dimanche"),
        "cuisine_par_repas_min": max(5, t // 3),
        "strategie": ("batch cooking + assemblage rapide" if batch
                      else "cuisine quotidienne à la demande"),
    }
